File: src/infrastructure/repository/builder.py
from typing import Type
from domain.abstracts.base_domain import BaseDomain
from infrastructure.repository.strategies.abstracts.base_repository_strategy import BaseRepositoryStrategy
from infrastructure.servers.abstracts.base_server import BaseServer

class Builder(BaseRepositoryStrategy):

    def __init__(self, server: BaseServer, domain_cls: Type[BaseDomain]):
        self._servers = server
        self._domain_cls = domain_cls

    def build_domain(self) -> BaseDomain:
        return self._domain_cls(
            client=self._servers,
            model_name=self._domain_cls.get_domain_name()
        )

File: angular-app/src/app/ddd-core/infrastructure/repository/builder.ts
import { DomainConstructor } from '../../application/strategies/abstracts/base-application-strategy';
import { BaseDomain } from '../../domain/abstracts/base-domain';
import { BaseServer } from '../servers/abstracts/base-server';

export class Builder {
  private _server: BaseServer | null = null;

  constructor(
    private readonly domainClass: DomainConstructor,
    private readonly serverFactory?: () => BaseServer,
    serverInstance: BaseServer | null = null
  ) {
    this._server = serverInstance;
  }

  get server(): BaseServer {
    if (!this._server) {
      if (!this.serverFactory) {
        throw new Error('Builder: serverFactory ou serverInstance deve ser fornecido');
      }
      this._server = this.serverFactory();
    }
    return this._server;
  }

  buildDomain(): BaseDomain {
    return new this.domainClass(this.server.loadServer(), this.domainClass.getDomainName());
  }
}
